java/ql/lib/semmle/code/java/dataflow/internal/DataFlowImpl.qll

```diff
@@ -82,6 +99,15 @@ abstract class Configuration extends string {
    */
   predicate isAdditionalFlowStep(Node node1, Node node2) { none() }
 
+  /**
+   * Holds if the additional flow step from `node1` to `node2` must be taken
+   * into account in the analysis. This step is only applicable in `state1` and
+   * updates the flow state to `state2`.
+   */
+  predicate isAdditionalFlowStep(Node node1, FlowState state1, Node node2, FlowState state2) {
```

Comment:
Won't we need something like this for cases where we previously chained together configurations?

```ql
/**
 * Holds if reaching `node1` updates the state from `state1` to `state2`.
 */
predicate updatesState(Node node, FlowState state1, FlowState state2)
```

Reply:
You can simulate that fairly well by using `isAdditionalStep(node, state1, node, state2)`. So I guess we could add it as a convenience, but it shouldn't be strictly necessary.

Reply:
Except that will only work for taint steps and not value preserving steps.

Reply:
I wasn't planning on allowing state-dependent or state-changing value steps. Do you think we'll have strong use-cases for that?

Reply:
Not sure, but I do know that we have previously chained together configurations to model the case where flow must go through a certain node. Whether the step into that node is alway a taint step I don't know, but it seems a bit restrictive to assume so.

Reply:
I think I've found a use case for state-changing value steps when there's a multi-part barrier (in this case, two different methods that need to be called to secure an XML parser against XXE vulnerabilities).